### PR TITLE
Typescriptize DataTable component and more

### PR DESCRIPTION
**Pull request by magicznyleszek**
## Checklist

1. [ ] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [ ] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes

## Description

Internal improvements and cleanups. Some unnoticed unknown bugs (mainly of "something something is `undefined`" type) might've been fixed.

## Notes

The main gist of this is moving `DataTable` to TypeScript. Each file imported into it was also typescriptized.

I've touched many files here, most changes are:
- Typescriptized `stores.pageState` and moved to separate file (this touched a lot of files, but each in a small way)
  - Removed unused `shouldShowNewFeatureDialog` from `FormLanding` component
- Further improved types in the `actions.d.ts` file (the ones touched by `DataTable` typescriptization) and in submissions actions
- Made `onChange` optional in `Checkbox` component, so that we can use it through `onClick` only
- Changed `Icon` into functional component
- Fixed `placement` prop in `KoboDropdown` to be optional, as it has default value
- Rewrote (to simplify and encapsulate) types and constants responsible for handling validation statuses: `validationStatus.constants.ts`
- Typescriptized `ValidationStatusDropdown` component
- Typescriptized `SubmissionModal` component
  - It's the one responsible for that big Single Submission Modal
  - Had to modify many small bits, because TypeScript found many issues with what types of variables are we joggling around in here
  - The changes are more severe in this big file, because of validation statuses code changes - unfortunately GitHub doesn't is not able to display a nicer diff here
  - Moved its styles into separate file from `modal.scss` and cleaned them up a bit
- `AudioCell` component now handles error received through `mediaAttachment` property
- `MediaCell` component now handles error received through `mediaAttachment` property
- `TableMediaPreview` component now handles error received through `mediaAttachment` property
- Typescriptized the monstrosity of `DataTable` component
  - All the other typescriptized components are being used here, thus the need to change them
  - Validation statuses code changes also take some role in the amount of changes
  - Crated a new separate file just for types: `table.types.ts`
  - Updated `react-table` to newer version (not a big jump, just wanted latest bugfixes for our version)
  - Added `@types/react-table`
  - Modified how we handle shift key selection (deprecated `window.event` was used, now we switched to getting it from `onClick`) → this caused the small change in `Checkbox`
  - Now all these mysterious "instances" in the code make sense (I hope, if not I will add some comments)
- Typescriptized `ColumnsHideDropdown` component
- Typescriptized `ColumnsHideForm` component and encapsulated better getting the fuse search results
- Typescriptized `TableBulkCheckbox` component
- Typescriptized `TableBulkOptions` component
- Typescriptized `TableColumnSortDropdown` component
- In `tableStore` I did some cleanups and started using property names from constants instead of literal strings
- Some improvements in `dataInterface` types 
- Typescriptized `enketoHandler`

#### Testing

I know it's quite big, so the main areas to look at are:
- Clicking around everything in Data Table
- Checking all the places that modify validation status:
  - Data Table column header filtering
  - Data Table cell validation status changing
  - "Change status" above Data Table when multiple checkboxes are selected
  - Single Submission modal has dropdown for changing status
- Testing Data Table cells with media type responses (audio, video), also with broken files (e.g. rename `test.txt` into `test.mp3` and upload as `audio` response :smile:)
- Testing opening enketo - "Edit" and "View" buttons in Single Submission modal

### CODE REVIEW

File: jsapp/js/components/submissions/table.tsx [hunk 1]
```diff
@@ -555,10 +621,10 @@ export class DataTable extends React.Component {
             asset={this.props.asset}
             fieldId={VALIDATION_STATUS_ID_PROP}
             sortValue={tableStore.getFieldSortValue(VALIDATION_STATUS_ID_PROP)}
-            onSortChange={this.onFieldSortChange}
+            onSortChange={this.onFieldSortChange.bind(this)}
             onHide={this.onHideField}
```
Comment (jamesrkiger) [inline jsapp/js/components/submissions/table.tsx:625]: I think this is missing `.bind(this)`

**Comment by jamesrkiger** [comment]
The full screen toggle and settings buttons here are broken. I think the function calls are missing `.bind(this)`
image

File: jsapp/js/components/formLanding.js [hunk 2]
```diff
@@ -46,7 +46,6 @@ class FormLanding extends React.Component {
       nextPagesVersions: [],
       anonymousSubmissions: false,
       anonymousPermissions: [],
-      shouldShowNewFeatureDialog: false,
```
Comment (jamesrkiger) [inline jsapp/js/components/formLanding.js:49]: Just wanted to highlight this to make sure you're removing it on purpose. I assume so, but wanted to check.

**Comment by jamesrkiger** [comment]
Looks like the filter here is broken. It doesn't do anything when I type in that field:
image

File: jsapp/js/components/formLanding.js (hunk 2 above)
Comment (magicznyleszek) [reply jsapp/js/components/formLanding.js:49]: Yeah, I removed it because there is no place that used it

**Comment by magicznyleszek** [comment]
@jamesrkiger Fixed missing `.bind`s, filter modal, and the fullscreen/settings buttons.

**Pull request merged**